numeric_ranges_algorithms/numeric_ranges_algorithms.md

```diff
@@ -0,0 +1,867 @@
+
+3. P3179R8 requires random access ranges (see Section 2.6).
+
+Of these differences, (1) and (2) could apply generally to all ranges algorithms, so we adopt them for this proposal.
+
+Regarding (1), this would make our proposal the first to add non-parallel range-as-output ranges algorithms to the Standard.  For arguments in favor of non-parallel algorithms taking a range as output, please refer to <a href="https://wg21.link/P3490R0">P3490R0</a>, "Justification for ranges as the output of parallel range algorithms."  (Despite the title, it has things to say about non-parallel algorithms too.)  Taking a range as output would prevent use of existing output-only iterators that do not have a separate sized sentinel type, like `std::back_insert_iterator`.  However, all the algorithms we propose require at least forward iterators (see below).  P3490R0 shows that it is possible for both iterator-as-output and range-as-output overloads to coexist, so we follow P3179R8 by not proposing iterator-as-output algorithms here.
+
+Regarding (2), we make the parallel algorithms proposed here take sized random access ranges, as P3179R8 does.  For consistency, we also propose that the output ranges be sized.  As a result, any parallel algorithms with an output range need to return both an iterator to one past the last element of the input, and an iterator to one past the last element of the input.  This tells callers whether there was enough room in the output, and if not, where to start when processing the rest of the input.  This includes all the `*{ex,in}clusive_scan` algorithms we propose.
+
+Difference (3) relates to P3179R8 only proposing parallel algorithms.  It would make sense for us to relax this requirement for the non-parallel algorithms we propose.  This leaves us with two possibilities:
+
+* (single-pass) input and output ranges, the most general; or
+
+* (multipass) forward ranges.
+
+The various reduction and scan algorithms we propose can combine the elements of the range in any order.  For this reason, we make the non-parallel algorithms take (multipass) forward ranges, even though this is not consistent with the existing non-parallel `<numeric>` algorithms.  If users have single-pass iterators, they should just call one of the `fold_*` algorithms, or use the `views::scan` proposed elsewhere.  This has the benefit of letting us specify `ranges::reduce` to return just the value.  We don't propose a separate algorithm `reduce_with_iter`, as we explain elsewhere in this proposal.
+
+## Constexpr parallel algorithms?
+
+<a href="https://wg21.link/p2902r1">P2902R1</a> proposes to add `constexpr` to the parallel algorithms.  P3179R8 does not object to this; see Section 2.10.  We continue the approach of P3179R8 in not opposing P2902R1's approach, but also not depending on it.
+
+## `ranges::reduce` design
+
+In this section, we focus on `ranges::reduce`'s design.  The discussion here applies generally to the other algorithms we propose.

```

Comment:
What we could do to address the issue with unknown identity operations for `reduce` is to suggest overloads that take an optional identity value, or suggest a protocol (a "trait" or some such) to extract that value out of the reduction operation itself.

I am not saying we should do that. But I see it as an API gap preventing more efficient parallel implementations of reduction, especially with  the `unseq` policy.

Reply:
> The various *reduce algorithms should either always require an identity value (recommended in the various ranges plan papers) or should use T{} as the default identity value. Regardless, it's up to users to define the identity value.

This sounds describing the initial value for the algorithm, not the identity value of the operation. These are not the same, let's try disambiguating those. Above, I considered an identity value parameter, additional to the initial value parameter.